README
Gentoo::Util::VirtualDepend

This module serves as a low level glue layer for the handful of manual
mappings that are needed in Gentoo due to things not strictly tracking
upstream.

"CPANPLUS::Dist::Gentoo" has similar logic to this, but not as simple (
or for that matter, usable without "CPANPLUS" )

This module is not intended to be used entirely on its own, but as a
short-circuit before calling complicated "MetaCPAN" code.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Gentoo-Util-VirtualDepend

0.003024 2017-03-19T05:22:36Z 4f6f1ae

0.003023 2017-03-19T05:21:07Z 4385424
 - Remove special casing for Gtk2

0.003022 2017-03-06T13:07:00Z c56b6cd
 [Bugfix]
 - Update tests for -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.003021, see misc/*.deps* for details
 - develop: ↑1 (suggests: ↑2)
 - test: ↓1

0.003021 2017-01-27T12:24:51Z 597553e
 - Update metadata map

0.003020 2016-12-10T18:49:40Z 5e1d1aa
 - Remove special casing for ExtUtils::PkgConfig

 [Dependencies::Stats]
 - Dependencies changed since 0.003019, see misc/*.deps* for details
 - develop: +1

0.003019 2016-09-12T17:55:27Z e94b404
 - Remove special casing for Data::Buffer
 - Remove DBIx::SearchBuilder special case
 - Remove Digest::BubbleBabble special case
 - Remove Digest::MD2 special case

0.003018 2016-08-27T13:51:07Z 3fe5211
 - Remove Cache::Mmap special case.
 - When more than one module reports to index a given package, prefer the one that matches the module name.
 - Add Pod-Perldoc to virtuals list due to being impled in dev-lang/perl, even though the virtual does not itself
   exist.

 [Utils / vdep_check]
 - now uses Corelist directly for differencing instead of shelling out. This is much faster as
   Gentoo::Util::VirtualDepend loads Corelist anyway for its override checks.
 - now reports both changed and unchanged dists
 - now reports inconsistently versioned dists and inconsistently changed dists

 [Utils / vdep_report]
 - new utility, dumps an entire Perl's corelist grouped by Gentoo classification.

0.003017 2016-08-03T18:26:16Z eaa81e5
 - Remove special casing for: Regexp::Common

 [Dependencies::Stats]
 - Dependencies changed since 0.003016, see misc/*.deps* for details
 - develop: (suggests: ↑1)

0.003016 2016-07-26T09:47:48Z 87d9e6b
 - Remove special casing for: Class::Loader, Convert::ASCII::Armour, Convert::PEM

0.003015 2016-04-30T21:41:26Z 7badff5
 - Remove special casing for: Text::Template, Text::Wrapper, Tie::EncryptedHash, YAML

0.003014 2016-04-25T14:45:19Z aca0ad3
 - Remove Virtual rules for dual-life things that we don't have virtuals for yet. This means they will fall back to the
   corelist resolution strategy when seen. They will be re-added at such time as they are actually needed.

0.003013 2016-04-18T18:47:46Z af611c6
 - Remove Wx and DateManip special casing
 - Remove Tk, MogileFS-Server and Ogg-Vorbis-Header special casing
 - Update mappings
 - Remove Math-Pari special casing.
 - Remove PerlTidy special casing.
 - Remove Set-Scalar special casing.
 - Remove String-CRC32 special casing.
 - Remove Time::Period special casing.

0.003012 2016-03-10T07:51:02Z 8020cc9
 - Update mappings.
 - Remove special casing for Crypt-* family
 - Remove special casing for Net-*

0.003011 2016-03-01T13:43:31Z 0a3ca04
 - Remove special casing for CGI-SpeedyCGI and Locale-Maketext-Fuzzy

0.003010 2016-02-27T23:03:24Z fe322c7
 - Upgrade Minimum Assumed Supported Perl to 5.18.2 (history/f1379d9)
 - Upgrade Maximum Assumed Supported Perl to 5.22.1 (master/6a01824)

 [Dependencies::Stats]
 - Dependencies changed since 0.003007, see misc/*.deps* for details
 - runtime: ↑1

0.003007 2016-02-27T21:02:56Z fad2c0e
 [Mappings]
 - Update to latest versions
 - Remove special casing for Image-Info and Image-Size

0.003006 2016-01-09T19:29:01Z e64429c
 [Dependencies::Stats]
 - Dependencies changed since 0.003005, see misc/*.deps* for details
 - develop: (suggests: ↑1)

 [Mappings]
 - Remove special casings for Module-Info and txt2html

0.003005 2015-10-08T07:27:21Z 855094f
 [Metadata]
 - Sync static data from metacpan queries

0.003004 2015-09-13T04:43:40Z d357b13
 [Dependencies::Stats]
 - Dependencies changed since 0.003003, see misc/*.deps* for details
 - develop: +2 ↑1 (suggests: ↑1)

 [Mappings]
 - Remove special casings for Text-AutoFormat and Text-Reform

0.003003 2015-08-04T09:57:15Z 759505d
 [Mappings]
 - Remove special casings for Gtk2Fu: Has gone to CPAN nirvana.
 - Actually remove special casing for Inline-Files

0.003002 2015-07-18T18:50:45Z ff0171b
 [Dependencies::Stats]
 - Dependencies changed since 0.003001, see misc/*.deps* for details
 - develop: +3 -1 (suggests: ↑1)

 [Mappings]
 - Remove special casing for Locale-MakeText-Lexicon

0.003001 2015-06-21T20:54:50Z ac103ec
 [Mappings]
 - Drop CGI and Module-Build special cases

0.003000 2015-06-02T18:43:52Z 57a37f6
 [Features]
 - Added get_known(gentoo_packages|dists|modules) functions.
 - Added resolver map showing source data used to aggregate data.

0.002005 2015-05-12T10:10:10Z b8ed3de
 [Dependencies::Stats]
 - Dependencies changed since 0.002004, see misc/*.deps* for details
 - develop: (suggests: ↑1)

 [Mappings]
 - Drop specialcasing for Class-ReturnValue

0.002004 2015-05-01T17:33:29Z 7ecae13
 [Mappings]
 - Drop need for specialcasing for ExtUtils-Depends
 - Add entry for CGI::HTML::Functions -> CGI

0.002003 2015-04-19T03:42:39Z 3ec3162
 [Mappings]
 - Dropping need for specialcasings for CGI-Simple
 - Adding new Encode::XS modules.

0.002002 2015-04-06T21:02:20Z 5e81d82
 [Mappings]
 - Dropping need for specialcasings for Log-Dispatch

0.002001 2015-04-02T01:50:43Z 79ff0fe
 [Mappings]
 - Dropped need for specialcasing Config-General

0.002000 2015-04-01T16:52:44Z 8de5a28
 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: +1 (recommends: +1 ↑1)
 - develop: +6 ↑2 -2 (recommends: +1, suggests: ↑2)
 - runtime: +2 ↓1 -1 (recommends: +2)
 - test: +1 (recommends: ↑1 ↓1)

 [Features]
 - Add a package->contents mapping ( mostly for consistency checking )
 - add a module_is_perl function that checks corelist

 [Mappings]
 - Synced mappings to current CPAN/tree status

0.001000 2014-10-10T17:53:57Z 9a68f55
 - First version.


util/vdep_check.pl
#!/usr/bin/env perl
# FILENAME: vdep_check.pl
# CREATED: 06/02/15 12:43:42 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Show result of CPAN version deltas in terms of virtuals.

use 5.020;
use warnings;

use Capture::Tiny qw(capture_stdout);
use Data::Dump qw( pp );
use Gentoo::Util::VirtualDepend;
use List::UtilsBy qw( uniq_by );
use Gentoo::PerlMod::Version qw( gentooize_version );
my $v = Gentoo::Util::VirtualDepend->new();

my ( $lines, ) = capture_stdout {
  system( 'corelist', '--diff', '5.20.2', '5.22.0' );
};

my $table = {};

for my $line ( split /\n/, $lines ) {
  my ( $module, $old, $new ) = $line =~ /\A(\S+)\s+(\S+)\s+(\S+)\s*\z/;

  # these versions for the sake of upgrade testing don't exist
  next if $old eq '(absent)' and $new eq '(undef)';

  my $type;
  if ( $new eq '(absent)' ) {
    $type = 'REMOVE';
  }
  elsif ( $old eq '(absent)' ) {
    $type = "ADD";
  }
  else {
    $type = "UPGRADE";
  }
  my $override = "untracked";
  if ( $v->has_module_override($module) ) {
    $override = $v->get_module_override($module);
  }
  $type = simplify( [ $type, $old, $new ] ) if $override eq 'untracked';
  $table->{$override} //= {};
  $table->{$override}->{$module} = [ $type, $old, $new ];
}

for my $key ( keys %$table ) {
  next if $key eq 'untracked';
  my $entries = $table->{$key};
  my (@ev) = values %{$entries};
  my (@types) = uniq_by { $_->[0] } @ev;

  if ( 1 == @types && $types[0]->[0] eq 'REMOVE' ) {
    $table->{$key} = ['REMOVE'];
    next;
  }
  my (@vnew) = uniq_by { $_->[2] } @ev;
  my (@vold) = uniq_by { $_->[1] } @ev;

  if ( 1 == @vnew and 1 == @vold ) {
    $table->{$key} = [ "UPGRADE", $vold[0]->[1], $vnew[0]->[2] ];
    next;
  }
  elsif ( 1 == @vnew ) {
    $table->{$key} = [ "UPGRADE", "mixed", $vnew[0]->[2] ];
    next;
  }
}

for my $key ( sort keys %$table ) {
  next if $key eq 'untracked';
  if ( ref $table->{$key} eq 'ARRAY' ) {
    if ( $table->{$key}->[0] eq 'REMOVE' ) {
      printf "%-40sremoved from perl\n", $key;
      next;
    }
    if ( $table->{$key}->[0] eq 'ADD' ) {
      my $v = gentooize_version( $table->{$key}->[1], { lax => 1 } );
      printf "%-40s+%-30s\t\t[=%s, new distribution+virtual in perl]\n", $key, $v, $table->{$key}->[1];
    }
    if ( $table->{$key}->[0] eq 'UPGRADE' ) {
      my $v = gentooize_version( $table->{$key}->[2], { lax => 1 } );
      printf "%-40s+%-30s\t\t[=%s, was %s]\n", $key, $v, $table->{$key}->[2], $table->{$key}->[1];
      next;
    }
    if ( $table->{$key}->[0] eq 'NOCHANGE' ) {
      printf "%-40sno change between perls\n", $key;
      next;
    }
  }
  printf "%s:\n---\t\t%s\n", $key, pp $table->{$key};
}

sub simplify {
  my ($node) = @_;
  return $node->[0] if $node->[0] eq 'REMOVE';
  return sprintf '%s: %3$s',         @{$node} if $node->[0] eq 'ADD';
  return sprintf '%s: %2$s => %3$s', @{$node} if $node->[0] eq 'UPGRADE';
  die "What is $node?";
}


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw(
  de
  Hah
));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



maint/update-dist-to-gentoo.pl
#!/usr/bin/env perl
# FILENAME: update-dist-to-gentoo.pl
# CREATED: 10/11/14 03:21:18 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Update dist-to-gentoo file.

use strict;
use warnings;
use utf8;

my %premap = ();

my (@normal_virtuals) = qw(
  Archive-Tar
  Attribute-Handlers
  AutoLoader
  autodie
  B-Debug
  CPAN
  CPAN-Meta
  CPAN-Meta-Requirements
  CPAN-Meta-YAML
  Carp
  Compress-Raw-Bzip2
  Compress-Raw-Zlib
  DB_File
  Data-Dumper
  Devel-PPPort
  Digest
  Digest-MD5
  Digest-SHA
  Dumpvalue
  Encode
  Exporter
  ExtUtils-CBuilder
  ExtUtils-Command
  ExtUtils-Constant
  ExtUtils-Install
  ExtUtils-MakeMaker
  ExtUtils-Manifest
  ExtUtils-ParseXS
  File-Path
  File-Temp
  Filter-Simple
  Getopt-Long
  HTTP-Tiny
  I18N-LangTags
  IPC-Cmd
  IO
  IO-Compress
  IO-Zlib
  IO-Socket-IP
  JSON-PP
  Locale-Maketext
  Locale-Maketext-Simple
  MIME-Base64
  Math-BigInt
  Math-BigInt-FastCalc
  Math-BigRat
  Math-Complex
  Memoize
  Module-CoreList
  Module-Load
  Module-Load-Conditional
  Module-Loaded
  Module-Metadata
  Net-Ping
  Package-Constants
  Params-Check
  Parse-CPAN-Meta
  Perl-OSType
  Pod-Escapes
  Pod-Parser
  Pod-Perldoc
  Pod-Simple
  Safe
  Scalar-List-Utils
  Socket
  Storable
  Sys-Syslog
  Term-ANSIColor
  Term-ReadLine
  Test
  Test-Harness
  Test-Simple
  Text-Balanced
  Text-ParseWords
  Text-Tabs+Wrap
  Thread-Queue
  Thread-Semaphore
  Tie-RefHash
  Time-HiRes
  Time-Local
  Time-Piece
  XSLoader
  autodie
  bignum
  if
  libnet
  parent
  podlators
  threads
  threads-shared
  version
  Unicode-Collate
  Unicode-Normalize
);

for my $normal (@normal_virtuals) {
  $premap{$normal} = 'virtual/perl-' . $normal;
}
use Data::Handle;
my $handle = Data::Handle->new('main');

use FindBin;
use Path::Tiny qw(path);

while ( my $line = <$handle> ) {
  chomp $line;
  my ( $key, $value ) = split /,/, $line;
  if ( not $key ) {
    warn "> $line ";
    next;
  }
  $premap{$key} = $value;
}

my $target = path($FindBin::Bin)->sibling('share')->child('dist-to-gentoo.csv');
my $fh     = $target->openw_raw;
for my $key ( sort keys %premap ) {
  $fh->printf( "%s,%s\n", $key, $premap{$key} );
}

package main;

__DATA__
AcePerl,dev-perl/Ace
App-SVN-Bisect,dev-util/App-SVN-Bisect
Autodia,dev-util/autodia
BioPerl,sci-biology/bioperl
BioPerl-DB,sci-biology/bioperl-db
BioPerl-Network,sci-biology/bioperl-network
BioPerl-Run,sci-biology/bioperl-run
Frontier-RPC,dev-perl/frontier-rpc
GBrowse,sci-biology/GBrowse
Glib,dev-perl/glib-perl
Gnome2,dev-perl/gnome2-perl
Gnome2-Canvas,dev-perl/gnome2-canvas
Gnome2-VFS,dev-perl/gnome2-vfs-perl
Gnome2-Wnck,dev-perl/gnome2-wnck
Gtk2,dev-perl/gtk2-perl
Gtk2-Ex-FormFactory,dev-perl/gtk2-ex-formfactory
Gtk2-GladeXML,dev-perl/gtk2-gladexml
Gtk2-Spell,dev-perl/gtk2-spell
Gtk2-TrayIcon,dev-perl/gtk2-trayicon
Gtk2-TrayManager,dev-perl/gtk2-traymanager
Image-ExifTool,media-libs/exiftool
NTLM,dev-perl/Authen-NTLM
OLE-Storage_Lite,dev-perl/OLE-StorageLite
Padre,app-editors/padre
PathTools,virtual/perl-File-Spec
Snapback2,app-backup/snapback2
XML-XSH2,app-editors/XML-XSH2
ack,sys-apps/ack
gettext,dev-perl/Locale-gettext


lib/Gentoo/Util/VirtualDepend.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Util::VirtualDepend;

our $VERSION = '0.003024';

# ABSTRACT: Hard-coded replacements for perl-core/ dependencies and dependencies with odd names in Gentoo

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );
use Path::Tiny qw( path );
use File::ShareDir qw( dist_file );

# Note: this should be the version default max_perl is in
use Module::CoreList 5.20151213;










# Note: This should be the latest visible version in portage at time of release
has max_perl => ( is => 'ro', lazy => 1, default => sub { '5.22.1' } );










# Note: This should be the lowest visible version 12 months prior to the time of release
has min_perl => ( is => 'ro', lazy => 1, default => sub { '5.18.2' } );

my %MOD2GENTOO;
my $MOD2GENTOO_LOADED;
my $MOD2GENTOO_FILE = 'module-to-gentoo.csv';

my %DIST2GENTOO;
my $DIST2GENTOO_LOADED;
my $DIST2GENTOO_FILE = 'dist-to-gentoo.csv';

my %GENTOO2DIST;
my %GENTOO2MOD;

my $DIST = q[Gentoo-Util-VirtualDepend];

sub _load_mod2gentoo {
  return if $MOD2GENTOO_LOADED;
  my $fh = path( dist_file( $DIST, $MOD2GENTOO_FILE ) )->openr_raw;
  while ( my $line = <$fh> ) {
    chomp $line;
    my ( $module, $map ) = split /,/, $line;    ## no critic (RegularExpressions)
    $MOD2GENTOO{$module} = $map;
    $GENTOO2MOD{$map} = [] unless exists $GENTOO2MOD{$map};
    push @{ $GENTOO2MOD{$map} }, $module;
  }
  return $MOD2GENTOO_LOADED = 1;
}

sub _load_dist2gentoo {
  return if $DIST2GENTOO_LOADED;
  my $fh = path( dist_file( $DIST, $DIST2GENTOO_FILE ) )->openr_raw;
  while ( my $line = <$fh> ) {
    chomp $line;
    my ( $module, $map ) = split /,/, $line;    ## no critic (RegularExpressions)
    $DIST2GENTOO{$module} = $map;
    $GENTOO2DIST{$map} = [] unless exists $GENTOO2DIST{$map};
    push @{ $GENTOO2DIST{$map} }, $module;
  }
  return $DIST2GENTOO_LOADED = 1;
}

sub has_module_override {
  my ( undef, $module ) = @_;
  _load_mod2gentoo unless $MOD2GENTOO_LOADED;
  return exists $MOD2GENTOO{$module};
}

sub get_module_override {
  my ( undef, $module ) = @_;
  _load_mod2gentoo unless $MOD2GENTOO_LOADED;
  return $MOD2GENTOO{$module};
}

sub has_dist_override {
  my ( undef, $dist ) = @_;
  _load_dist2gentoo unless $DIST2GENTOO_LOADED;
  return exists $DIST2GENTOO{$dist};
}

sub get_dist_override {
  my ( undef, $dist ) = @_;
  _load_mod2gentoo unless $DIST2GENTOO_LOADED;
  return $DIST2GENTOO{$dist};
}

sub has_gentoo_package {
  my ( undef, $package ) = @_;
  _load_dist2gentoo unless $DIST2GENTOO_LOADED;
  return exists $GENTOO2DIST{$package};
}

sub get_dists_in_gentoo_package {
  my ( undef, $package ) = @_;
  _load_dist2gentoo unless $DIST2GENTOO_LOADED;
  return @{ $GENTOO2DIST{$package} || [] };
}

sub get_modules_in_gentoo_package {
  my ( undef, $package ) = @_;
  _load_mod2gentoo unless $MOD2GENTOO_LOADED;
  return @{ $GENTOO2MOD{$package} || [] };
}

sub get_known_gentoo_packages {
  _load_dist2gentoo unless $DIST2GENTOO_LOADED;
  return keys %GENTOO2DIST;
}

sub get_known_dists {
  _load_dist2gentoo unless $DIST2GENTOO_LOADED;
  return keys %DIST2GENTOO;
}

sub get_known_modules {
  _load_mod2gentoo unless $MOD2GENTOO_LOADED;
  return keys %MOD2GENTOO;
}

sub module_is_perl {
  my ( $self, $opts, $module, $mod_version ) = @_;
  if ( not ref $opts ) {
    ( $opts, $module, $mod_version ) = ( {}, $opts, $module );
  }

  # If the module has a virtual, don't even consider
  # CPAN/perl decisions

  return if $self->has_module_override($module);
  require version;

  $opts->{min_perl} ||= $self->min_perl;
  my $min_perl = version->parse( $opts->{min_perl} );
  $opts->{max_perl} ||= $self->max_perl;
  my $max_perl = version->parse( $opts->{max_perl} );

  my $seen;
  ## no critic (Variables::ProhibitPackageVars)
  for my $version ( keys %Module::CoreList::version ) {
    my $perlver = version->parse($version);
    ## no critic (ControlStructures::ProhibitNegativeExpressionsInUnlessAndUntilConditions)
    next unless $perlver >= $min_perl;
    next unless $perlver <= $max_perl;

    # If any version in the range returns "deprecated", then we should
    # default to CPAN
    return if $Module::CoreList::deprecated{$version}{$module};

    # If any version in the range does not exist, then we should default to CPAN
    #
    return if not exists $Module::CoreList::version{$version}{$module};

    if ( not defined $mod_version ) {
      $seen = 1;
      next;
    }

    # If any version in the range is undef, and a specific version is requested,
    # Default to CPAN, because it means a virtual is not provisioned.
    return if not defined $Module::CoreList::version{$version}{$module};

    my $this_version = version->parse( $Module::CoreList::version{$version}{$module} );

    # If any version in the range is lower than required, default to CPAN
    # because it means a virtual is not provisioned, and a breakage will occur
    # on one of the versions.
    return if $this_version < version->parse($mod_version);

    $seen = 1;
  }
  return $seen;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::Util::VirtualDepend - Hard-coded replacements for perl-core/ dependencies and dependencies with odd names in Gentoo

=head1 VERSION

version 0.003024

=head1 SYNOPSIS

  use Gentoo::Util::VirtualDepend;

  my $v = Gentoo::Util::VirtualDepend->new();

  # somewhere in complex dependency resolution

  my $cpan_module = spooky_function();
  my $gentoo_dependency;

  if ( $v->has_module_override( $cpan_module ) ) {
    $gentoo_dependency = $v->get_module_override( $cpan_module );
  } else {
    # do it the hard way.
  }

If you're trying to be defensive and you're going to map the modules to distributions
the hard way ( trust me, the code is really ugly ), then you may instead want

  if ( $v->has_dist_override( $cpan_dist ) ) {
    $gentoo_dependency = $v->get_dist_override( $cpan_dist );
  } else {
    # fallback to using dev-perl/Foo-Bar
  }

Which basically serves as a distribution name translator.

=head2 WHY YOU WANT TO DO THAT

Well ...

     { requires => { Foo => 1.0 }}

     Foo is in Bar

     Foo 1.0 could have been shipped in in Bar-1.0, Bar-0.5, or Bar-2.0 for all you know.

That's the unfortunate reality of C<CPAN> dependencies.

So if you naively map

    Foo-1.0 → >=dev-lang/Bar-1.0

You might get breakage if C<Foo 1.0> didn't ship till C<Bar-2.0>, and the user has C<Bar-1.0> → Shower of sparks.

=head1 DESCRIPTION

This module serves as a low level glue layer for the handful of manual mappings
that are needed in Gentoo due to things not strictly tracking upstream.

C<CPANPLUS::Dist::Gentoo> has similar logic to this, but not as simple ( or for that matter, usable without C<CPANPLUS> )

This module is not intended to be used entirely on its own, but as a short-circuit before calling
complicated C<MetaCPAN> code.

=head1 METHODS

=head2 has_module_override

  $v->has_module_override( $module )

Returns true if there is a known mapping for C<$module> in C<Gentoo> that is unusual and may require translation.

Will return true for anything that is either a C<virtual> or has an unusual
name translation separating it from C<CPAN>.

=head2 get_module_override

  $v->get_module_override( $module )

Returns a C<Gentoo> dependency atom corresponding to C<$module> if there is a known mapping for C<$module>.

For instance,

  $v->get_module_override('ExtUtils::MakeMaker')

Emits:

  virtual/perl-ExtUtilsMakeMaker

If C<ExtUtils::MakeMaker> is one day de-cored (Hah!, dreams are free) then
C<has_module_override> will return false, and that instructs you to go back
to assuming it is in C<dev-perl/>

=head2 has_dist_override

  $v->has_dist_override( $distname )

Similar to C<has_module_override> but closer to the dependency spec.

Will return true for anything that is either a C<virtual> or has an unusual
name translation separating it from C<CPAN>.

=head2 get_dist_override

  $v->get_dist_override( $distname )

Similar to C<get_module_override> but closer to the dependency spec.

For instance:

  $v->get_dist_override('PathTools')

Emits:

  virtual/perl-File-Spec

Because C<Gentoo> is quirky like that.

=head2 has_gentoo_package

  $v->has_gentoo_package( 'virtual/perl-Test-Simple' )

Determines if the data file has entries mapping to C<virtual/perl-Test-Simple>.

This is mostly for internal consistency tests/maintenance.

=head2 get_dists_in_gentoo_package

  my @list = $v->get_dists_in_gentoo_package( 'virtual/perl-Test-Simple' )

Returns a list of C<CPAN> Distributions that map to this dependency.

=head2 get_modules_in_gentoo_package

  my @list = $v->get_modules_in_gentoo_package( 'virtua/perl-Test-Simple' )

Returns a list of modules that map to this dependency.

=head2 get_known_gentoo_packages

  my @list = $v->get_known_gentoo_packages

Returns a list of Gentoo packages for which there are known overrides.

=head2 get_known_dists

  my @list = $v->get_known_dists

Returns a list of C<CPAN> Distributions for which there are known overrides

=head2 get_known_modules

  my @list = $v->get_known_modules

Return a list of C<CPAN> Modules for which there are known overrides

=head2 module_is_perl

This function determines if it is "safe" to assume availability
of a given module ( or a given module and version ) without needing to
stipulate either a virtual or a C<CPAN> dependency.

  ->module_is_perl( $module )
  ->module_is_perl( $module, $min_version )
  ->module_is_perl( \%config, $module, $min_version )

Rules:

=over 4

=item * If the module is present in the override map, then it is deemed B<NOT>
available from C<Perl>, because you should be using the override instead.

=item * If the module is missing on any version in the range specified, then it is
B<NOT> available from C<Perl>, and you must depend on a virtual or some other
dependency you can source.

=item * If the module is marked I<deprecated> on any version in the range specified,
then it is assumed B<NOT> available in C<Perl> ( due to likely deprecation warnings
and imminent need to start adapting )

=item * If a minimum version is specified, and I<any> version of C<Perl> in the range
specified does not satisfy that minimum, then it is assumed B<NOT> available in
C<Perl> ( due to the inherent need to manually solve the issue via a virtual or a
minimum C<Perl> dependency )

=item * If a minimum version is specified, and I<any> version of C<Perl> in the range
specified is an explicit C<undef>, then it is assumed B<NOT> available in C<Perl>,
because clearly, one version of C<Perl> having C<undef> and another having an
explicit version, and needing only one of the two requires a manual dependency
resolution.

=back

Examples:

=over 4

=item * Determine if C<strict> is I<implicitly> available.

  if ( $v->module_is_perl( 'strict' ) ) {

=item * Determine if C<strict> version C<1.09> is available.

  if ( $v->module_is_perl( 'strict' => '1.09' ) ) {

This will of course return C<undef> unless C<min_perl> is at least C<5.21.7>.

Thus, if your support range is 5.18.0 to 5.20, and somebody stipulates that minimum,
you will have to declare a dependency on C<Perl> 5.21.7.

Even if your support range is 5.18.0 to 5.22.0, you will still have to declare a
dependency on 5.21.7 instead of assuming its presence.

=item * Determine if C<strict> version C<1.09> is available on X to Y C<Perls>.

For most code where the support range is fixed, this will be unnecessary,
and changing the defaults via C<< ->new( min_perl => ... , max_perl => ... ) >>
should be sufficient.

However:

  if( $v->module_is_perl( { min_perl => '5.21.7', max_perl => '5.21.9' }, 'strict', '1.09' ) ) {
      # true
  }

=back

=head1 ATTRIBUTES

=head2 max_perl

  ->new( max_perl => '5.20.2' )
  ->max_perl # 5.20.2

Stipulates the default maximum C<Perl> for L<< C<module_is_perl>|/module_is_perl >>.

=head2 min_perl

  ->new( min_perl => '5.20.2' )
  ->min_perl # 5.20.2

Stipulates the default minimum C<Perl> for L<< C<module_is_perl>|/module_is_perl >>.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Hard-coded replacements for perl-core/ dependencies and dependencies with odd names in Gentoo",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Gentoo-Util-VirtualDepend",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Gentoo::Util::VirtualDepend",
  "PREREQ_PM" => {
    "File::ShareDir" => 0,
    "Module::CoreList" => "5.20151213",
    "Moo" => 0,
    "Path::Tiny" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::File::ShareDir::Dist" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.003024",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir" => 0,
  "File::Spec" => 0,
  "Module::CoreList" => "5.20151213",
  "Moo" => 0,
  "Path::Tiny" => 0,
  "Test::File::ShareDir::Dist" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
